Inductor support for native c10d_functional

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113057
* #113215
* __->__ #112439

This PR adds Inductor support for [native c10d_functional ops](https://github.com/pytorch/pytorch/pull/110570).

The Inductor IRs introduced in this PR will replace the existing `CollectiveKernel` IR hierarchy. Compared to the existing collective IRs, the new IRs:
- Are target language agnostic and support AOTInductor.
- Express the constraints solely with read/write deps. This maximizes the potential for buffer reuse.
- Address an issue where out-of-place collective's input buffers could be mutated while being volatilely read.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

Commit:
Inductor support for native c10_functional

[ghstack-poisoned]

Commit:
Update on "Inductor support for native c10d_functional"


This PR adds Inductor support for [native c10d_functional ops](https://github.com/pytorch/pytorch/pull/110570).

The Inductor IRs introduced in this PR will replace the existing `CollectiveKernel` IR hierarchy. Compared to the existing collective IRs, the new IRs:
- Are target language agnostic and support AOTInductor.
- Express the constraints solely with read/write deps. This maximizes the potential for buffer reuse.
- Address an issue where out-of-place collective's input buffers could be mutated while being volatilely read.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Chillee commented on torch/_inductor/ir.py:
```diff
@@ -6686,6 +6689,202 @@ def codegen_collective(self, wrapper, output_name, input_names):
 
 
+# TODO(yifu): replace the CollectiveKernel IR hierarchy with _CollectiveKernel.
+class _CollectiveKernel(FallbackKernel):
```
In a subsequent PR will all of the existing CollectiveKernel stuff be removed?

yifuwang replied:
Yes

Chillee commented on torch/_inductor/ir.py:
```diff
@@ -6686,6 +6689,202 @@ def codegen_collective(self, wrapper, output_name, input_names):
+    @staticmethod
+    def wrap_inplace_input(inp):
+        return TensorBox.create(
+            InPlaceHint(
```
This is a separate discussion, but it was never clear to me the purpose of `InplaceHint`? Why is it needed?

yifuwang replied:
> This is a separate discussion, but it was never clear to me the purpose of InplaceHint?

@wconstab gave a nice explanation below. More concretely, during codegen, if the wrapper decides to reuse the input buffer for `InPlaceHint`, the latter effectively aliases the former. The reuse decision conveniently skip the allocation. Otherwise, a buffer is allocated for `InPlaceHint`, and the data of the input buffer is copied to the newly allocated buffer.

> Why is it needed?

This was to achieve oppurtunistic de-functionalization at IR level. It works, but I agree post grad passes are more suitable for this.

Chillee commented on torch/_inductor/lowering.py:
```diff
@@ -5119,6 +5119,82 @@ def all_to_all_single(
+    @register_lowering(_c10d_functional.all_reduce)
+    def _all_reduce(inp, reduce_op, group_name):
+        return ir.TensorBox.create(
+            ir._CollectiveKernel.create_inplace_single(
```
I don't quite understand these semantics. The underlying aten op is out of place right? Why can we just make it inplace here?

yifuwang replied:
> The underlying aten op is out of place right?

The underlying op is in-place (`all_reduce_`).

Chillee reviewed (commented):
I don't think I have a good understanding of the design of how collectives work in Inductor.

In particular, I'm not sure I understand the interaction between inplace and out-of-place things here.

When are things converted to inplace collectives?

Chillee commented:
Also cc: @yf225

Chillee commented:
I talked to @wanchaol a bit, and it sounds like the `InplaceHint` stuff is mostly used for inplacing. I think it's better to do this transformation at the FX level using the `reinplace_scatters` pass.

wconstab commented:
> I don't think I have a good understanding of the design of how collectives work in Inductor.
>
>In particular, I'm not sure I understand the interaction between inplace and out-of-place things here.
>
>When are things converted to inplace collectives?


In the original functional collectives, it was like this:  We trace things as functional + wait, purely so that its easier on the graph transformations.  But ultimately, we want to use inplace collective kernels in a lot of cases to save either a memcopy or an allocation.  so we only added an IR for functional, and we made inductor opportunistically do one of 2 things - (1) perform a copy of the input tensor into a fresh allocation and use an inplace c10d/nccl kernel on that allocation, (2) detect if it was safe to reuse an input buffer and if so, coordinate that with inductor's scheduler, and again, use the c10d/nccl inplace kernel.

I'm curious what the new design is too.  For sure there are things we could improve about how the above is coordinated with inductor.

Chillee commented:
Talking with @yifuwang, I think a better approach is to do what we’re doing with other ops we want to inplace, like scatters and custom Triton kernels.

Basically, if you have an inplace version, then it’s lowered as “clone + inplace op”. In many cases, we will already be able to remove the clone (such as if there is a pointwise op before the clone). However, we also have a pass to convert say, scatter into scatter_ in cases where it’s safe to do so.

The thing I’d like to avoid is having a somewhat separate system of Inplacing only for collectives.

Commit:
Update on "Inductor support for native c10d_functional"


This PR adds Inductor support for [native c10d_functional ops](https://github.com/pytorch/pytorch/pull/110570).

The Inductor IRs introduced in this PR will replace the existing `CollectiveKernel` IR hierarchy. Compared to the existing collective IRs, the new IRs:
- Are target language agnostic and support AOTInductor.
- Express the constraints solely with read/write deps. This maximizes the potential for buffer reuse.
- Address an issue where out-of-place collective's input buffers could be mutated while being volatilely read.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

yifuwang commented:
Updates:
- In-place `all_reduce` and `all_reduce_coalesced` with `reinplace_scatter` (renamed to `reinplace_inplaceable_ops`).

wconstab commented:
> The thing I’d like to avoid is having a somewhat separate system of Inplacing only for collectives.

agreed here

>  if you have an inplace version, then it’s lowered as “clone + inplace op”

the thing that's not clear to me is, we went to the trouble to convert usercode manually to using the 'functional collectives' so that we have functional ops in our graph.  So should we 
a) lower 'functional' collectives to inplace kernels?  (seems wrong, generally)
b) reinplace at the fx graph level after functionalization
c) something else? (is there a way to let inplace collectives be traced into the graph and not get functionalized? we didn't do the IR that way, but maybe we could change it)

Chillee commented:
>  lower 'functional' collectives to inplace kernels? 

I don't think there's anything wrong with this? We want functional operations throughout most of the compilation process since mutation makes passes much more difficult to write correctly. However, when it comes to the compiler and lowering, there's nothing wrong with us emitting functional collectives afterwards if it's more efficient.

> reinplace at the fx graph level after functionalization

Yes, this is currently the plan.

yifuwang commented:
@wconstab

> the thing that's not clear to me is, we went to the trouble to convert usercode manually to using the 'functional collectives' so that we have functional ops in our graph.

That's still valid effort. According to [Implicit Invariants for writing FX Graph Passes](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/README.md#implicit-invariants-for-writing-fx-graph-passes),

> After AOTDispatch tracing and before Inductor, we have no mutation in our graph, except for a copy_ epilogue at the end of the graph.

so we do want collectives to be functional thoughout this phase. Also,

> Additionally, we do have one pass that does introduce mutation - reinplace_scatters. This pass must run just before Inductor lowering, as otherwise this breaks our invariant.

This is how we introduce in-placing for other ops today. Since Inductor IR can express mutation there doesn't seem to be downsides.

Commit:
Update on "Inductor support for native c10d_functional"


This PR adds Inductor support for [native c10d_functional ops](https://github.com/pytorch/pytorch/pull/110570).

The Inductor IRs introduced in this PR will replace the existing `CollectiveKernel` IR hierarchy. Compared to the existing collective IRs, the new IRs:
- Are target language agnostic and support AOTInductor.
- Express the constraints solely with read/write deps. This maximizes the potential for buffer reuse.
- Address an issue where out-of-place collective's input buffers could be mutated while being volatilely read.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

yifuwang commented:
- Eliminate the usage of `MutationLayout` and handle mutation with `MutationOutput` introduced by @oulgen.
- Lowering `all_reduce`/`all_reduce_coalesced` as their in-place variant + `clone` even when they are not in-placeable.

oulgen commented:
Thanks for removing MutationLayout. I have killed all MutationLayout usages in ir.py on https://github.com/pytorch/pytorch/pull/112925

Only lowering.py usages left but they are not as simple

yifuwang commented:
To demonstrate that volatile reads of out-of-place collective inputs are handled properly, see the post-fusion diagram for an `all_gather_into_tensor_coalesced` example:

<img width="2371" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/a95c2cd8-7cbe-4f19-a214-15a99cc696cf">

`wait_tensor` has a dependency edge to the corresponding collective input. If the input were to be mutated, the mutation won't be scheduled between the collecitve kernel and the wait kernel.

oulgen commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -654,6 +655,8 @@ def reinplace_scatters(graph):
             mutated_inputs.add(node.args[0])
 
     def can_inplace(node, mutated_arg):
+        if isinstance(mutated_arg, (list, tuple)):
+            return all(can_inplace(node, arg) for arg in mutated_arg)
```
This is incorrect. Notice how can_inplace could `graph.erase_node(copy_node)`.

Chillee replied:
well, I do think it's not great that `can_inplace` does that :P

oulgen replied:
yeah, when inplace is not recursive, the solution is returning a ternary value. With this recursive version you would need ternary value with a list

yifuwang replied:
Good catch. The cases I'm dealing with never have a `copy_` epilogue so I overlooked that.

yifuwang replied:
@oulgen I think a cleaner way is to improve the overall handling of ops that mutates multiple inputs. However, that may cause too much conflict with your in-flight changes. I opted to simply move the copy epilogue removal out of `can_inplace` for now. Let me know what you think.

Commit:
Update on "Inductor support for native c10d_functional"


This PR adds Inductor support for [native c10d_functional ops](https://github.com/pytorch/pytorch/pull/110570).

The Inductor IRs introduced in this PR will replace the existing `CollectiveKernel` IR hierarchy. Compared to the existing collective IRs, the new IRs:
- Are target language agnostic and support AOTInductor.
- Express the constraints solely with read/write deps. This maximizes the potential for buffer reuse.
- Address an issue where out-of-place collective's input buffers could be mutated while being volatilely read.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

oulgen commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -686,12 +689,22 @@ def can_inplace(node, mutated_arg):
-            if can_inplace(node, node.args[inplaceable_op.mutated_arg]):
+            mutated_arg = node.args[inplaceable_op.mutated_arg]
+            if can_inplace(node, mutated_arg):
+                copy_node = copy_args_to_copy_nodes.get((mutated_arg, node))
```
you added support for mutated_arg possibly being a list, so wouldn't this be incorrect?

I think the only correct solution is to have a ternary return from can_inplace

Cannot_inplace, can_inplace no copy, can_inplace copy node to remove

oulgen replied:
Maybe handle the recursion out here and make can_inplace be a non recursive function?

yifuwang replied:
> you added support for mutated_arg possibly being a list, so wouldn't this be incorrect?

Hmm partially.

It is correct in the sense that list args in fx nodes are of type `torch.fx.immutable_collections.immutable_list`. They are hashable and can be used as dict keys.

It is incorrect because it doesn't properly remove copy epilogues for ops that mutates multiple args. Though I'm intentionally deferring this since (1) getting it right requires quite some changes and I'd like to do that after your improvements lands (2) we don't have such a case yet (we will soon once we add functionalization glue to all_reduce_).

oulgen replied:
Sounds good. Please add a note about how copy epilogue removal would only work for single argument. I'm hoping to land my improvements asap. 
https://github.com/pytorch/pytorch/pull/112801
It is already approved but the PR at the bottom of the stack is pending review

yifuwang replied:
Sounds good. Will add a note and will prioritize improving the handling of multi-mutation ops on top of your improvements once they land.

Commit:
Update on "Inductor support for native c10d_functional"


This PR adds Inductor support for [native c10d_functional ops](https://github.com/pytorch/pytorch/pull/110570).

The Inductor IRs introduced in this PR will replace the existing `CollectiveKernel` IR hierarchy. Compared to the existing collective IRs, the new IRs:
- Are target language agnostic and support AOTInductor.
- Express the constraints solely with read/write deps. This maximizes the potential for buffer reuse.
- Address an issue where out-of-place collective's input buffers could be mutated while being volatilely read.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

oulgen commented:
FYI my reinplace_pass improvements have landed. Happy to help you figure out how to adjust to my changes

Commit:
Update on "Inductor support for native c10d_functional"


This PR adds Inductor support for [native c10d_functional ops](https://github.com/pytorch/pytorch/pull/110570).

The Inductor IRs introduced in this PR will replace the existing `CollectiveKernel` IR hierarchy. Compared to the existing collective IRs, the new IRs:
- Are target language agnostic and support AOTInductor.
- Express the constraints solely with read/write deps. This maximizes the potential for buffer reuse.
- Address an issue where out-of-place collective's input buffers could be mutated while being volatilely read.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

LucasLLC commented on torch/_inductor/lowering.py:
```diff
@@ -5132,6 +5132,97 @@ def all_to_all_single(
         )
 
+    _c10d_functional = torch.ops._c10d_functional
+
```
Just noting we're about to land Broadcast here as well, so there may be some merge conflicts to handle -https://github.com/pytorch/pytorch/pull/112668

Let me know if I can help.

Commit:
Update on "Inductor support for native c10d_functional"


This PR adds Inductor support for [native c10d_functional ops](https://github.com/pytorch/pytorch/pull/110570).

The Inductor IRs introduced in this PR will replace the existing `CollectiveKernel` IR hierarchy. Compared to the existing collective IRs, the new IRs:
- Are target language agnostic and support AOTInductor.
- Express the constraints solely with read/write deps. This maximizes the potential for buffer reuse.
- Address an issue where out-of-place collective's input buffers could be mutated while being volatilely read.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Commit:
Update on "Inductor support for native c10d_functional"


This PR adds Inductor support for [native c10d_functional ops](https://github.com/pytorch/pytorch/pull/110570).

The Inductor IRs introduced in this PR will replace the existing `CollectiveKernel` IR hierarchy. Compared to the existing collective IRs, the new IRs:
- Are target language agnostic and support AOTInductor.
- Express the constraints solely with read/write deps. This maximizes the potential for buffer reuse.
- Address an issue where out-of-place collective's input buffers could be mutated while being volatilely read.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Commit:
Update on "Inductor support for native c10d_functional"


This PR adds Inductor support for [native c10d_functional ops](https://github.com/pytorch/pytorch/pull/110570).

The Inductor IRs introduced in this PR will replace the existing `CollectiveKernel` IR hierarchy. Compared to the existing collective IRs, the new IRs:
- Are target language agnostic and support AOTInductor.
- Express the constraints solely with read/write deps. This maximizes the potential for buffer reuse.
- Address an issue where out-of-place collective's input buffers could be mutated while being volatilely read.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge

Commit:
Update on "Inductor support for native c10d_functional"


This PR adds Inductor support for [native c10d_functional ops](https://github.com/pytorch/pytorch/pull/110570).

The Inductor IRs introduced in this PR will replace the existing `CollectiveKernel` IR hierarchy. Compared to the existing collective IRs, the new IRs:
- Are target language agnostic and support AOTInductor.
- Express the constraints solely with read/write deps. This maximizes the potential for buffer reuse.
- Address an issue where out-of-place collective's input buffers could be mutated while being volatilely read.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge
